zio-cli/shared/src/main/scala/zio/cli/completion/RegularLanguage.scala

```diff
@@ -0,0 +1,259 @@
+   */
+  def contains(tokens: List[String]): RIO[Has[CliConfig], Boolean] = {
+    val derivative = ZIO.foldLeft(tokens)(this)((language, word) => language.derive(word))
+    derivative.map(dx => dx.isNullable)
+  }
+
+  /**
+   * Returns a set consisting of the first token of all strings in this language
+   * that are useful for CLI tab completion. For infinite or unwieldly languages,
+   * it is perfectly fine to return the empty set: This will simply not display
+   * any completions to the user.
+   *
+   * If you'd like the cursor to advance to the next word when tab completion
+   * unambiguously matches the prefix to a token, append a space (" ") character
+   * to the end of the returned token. Otherwise, the cursor will skip to the
+   * end of the completed token in the terminal.
+   *
+   * Some examples of different use cases:
+   *  1. Completing file/directory names:
+   *     - Append a space to the ends of file names (e.g., "bippy.pdf"). This
+   *       is because we want the cursor to jump to the next argument position
+   *       if tab completion unambiguously succeeds.
+   *
+   *     - Do not append a space to the end of a directory name (e.g., "foo/").
+   *       This is because we want the user to be able to press tab again to
+   *       gradually complete a lengthy file path.
+   *
+   *     - Append a space to the ends of string tokens.
+   *
+   * You may be asking why we don't try to use the `-o nospace` setting of
+   * `compgen` and `complete`. The answer is they appear to be all or nothing:
+   * For a given tab completion execution, you have to choose one behavior or
+   * the other. This does not work well when completing both file names and
+   * directory names at the same time.
+   */
+  def firstTokens(prefix: String, compgen: Compgen): RIO[Blocking, Set[String]]
+}
+
+object RegularLanguage {
+
+  /**
+   * The `Empty` language (∅) accepts no strings.
+   */
+  case object Empty extends RegularLanguage {
+    def isNullable: Boolean = false
+
+    def derive(token: String) = ZIO.succeed(Empty)
+
+    def firstTokens(prefix: String, compgen: Compgen): RIO[Blocking, Set[String]] =
+      ZIO.succeed(Set.empty)
+
+    override def toString = "∅"
+  }
+
+  /**
+   * The `Epsilon` language (ε) accepts only the empty string.
+   */
+  case object Epsilon extends RegularLanguage {
+    def isNullable: Boolean = true
+
+    def derive(token: String) = ZIO.succeed(Empty)
+
+    def firstTokens(prefix: String, compgen: Compgen): RIO[Blocking, Set[String]] =
+      ZIO.succeed(Set(""))
+
+    override def toString = "ε"
+  }
+
+  sealed trait Token extends RegularLanguage {
+    def isNullable: Boolean = false
+  }
+
+  /**
+   * A `StringToken(value)` language represents the regular language that
+   * contains only `value`.
+   */
+  case class StringToken(value: String) extends Token {
+    def derive(token: String) =
+      ZIO
+        .access[Has[CliConfig]](cliConfig =>
```

Comment:
Since this pattern is used multiple times, it is worth adding a `CliConfig.cliConfig` that uses `ZIO.access[...]` to return the `CliConfig` in a ZIO effect.

Reply:
Makes sense, thanks. I _think_ I fixed this in an idiomatic way [here](https://github.com/zio/zio-cli/blob/98ce50ae7f7dcb4447da2f402468f4e980a09577/zio-cli/shared/src/main/scala/zio/cli/CliConfig.scala#L32). If that isn't quite what you'd like I will patch it up.